include/swift/AST/ArgumentList.h

```diff
@@ -0,0 +1,649 @@
+    }
+    iterator operator++(int) { // post-increment
+      auto tmp = *this;
+      ++*this;
+      return tmp;
+    }
+    iterator &operator--() { // pre-decrement
+      --Idx;
+      return *this;
+    }
+    iterator operator--(int) { // post-decrement
+      auto tmp = *this;
+      --*this;
+      return tmp;
+    }
+
+    iterator &operator+=(unsigned i) {
+      Idx += i;
+      return *this;
+    }
+    iterator operator+(unsigned i) const {
+      return iterator(ArgList, Idx + i);
+    }
+    iterator &operator-=(unsigned i) {
+      Idx -= i;
+      return *this;
+    }
+    iterator operator-(unsigned i) const {
+      return iterator(ArgList, Idx - i);
+    }
+
+    unsigned operator-(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx - rhs.Idx;
+    }
+    bool operator<(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx < rhs.Idx;
+    }
+    bool operator<=(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx <= rhs.Idx;
+    }
+    bool operator>(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx > rhs.Idx;
+    }
+    bool operator>=(const iterator &rhs) const {
+      assert(ArgList == rhs.ArgList);
+      return Idx >= rhs.Idx;
+    }
+  };
+
+  iterator begin() const { return {this, 0}; }
+  iterator end() const { return {this, size()}; }
+
+  Argument front() const { return get(0); }
+  Argument back() const { return get(size() - 1); }
+
+  Argument operator[](unsigned idx) const { return get(idx); }
+
+  /// Returns the index of the first trailing closure in the argument list, or
+  /// \c None if there are no trailing closures.
+  ///
+  /// Note that in type-checked AST, this may return an index of a variadic
+  /// expansion which contains the trailing closure (and it may not necessarily
+  /// be the first argument of the expansion).
+  Optional<unsigned> getRawFirstTrailingClosureIndex() const {
+    if (RawFirstTrailingClosureIndex == NumArgs)
+      return None;
+    return RawFirstTrailingClosureIndex;
+  }
+
+  /// Whether any trailing closures are present in the argument list.
+  bool hasAnyTrailingClosures() const {
+    return getRawFirstTrailingClosureIndex().hasValue();
+  }
+
+  /// Whether this argument list has a single trailing closure.
+  bool hasSingleTrailingClosure() const {
+    return hasAnyTrailingClosures() && !hasMultipleTrailingClosures();
+  }
+
+  /// Whether the argument list has multiple trailing closures.
+  bool hasMultipleTrailingClosures() const {
+    return HasMultipleTrailingClosures;
+  }
+
+  /// Whether a given index is for a labeled or unlabeled trailing closure.
+  ///
+  /// Note that in type-checked AST, this may return \c true for an index of a
+  /// variadic expansion which contains a trailing closure (and it may not
+  /// necessarily be the first argument of the expansion).
+  bool isRawTrailingClosureIndex(unsigned i) const {
```

Comment:
`isTrailingClosureArgument()`?

Reply:
Hmm, I think calling out the fact that it's a "raw" index is important, as it's not what you may expect in type-checked AST, in particular it might return true for default arguments or arguments that have been re-ordered after the trailing closure.

Alternatively we may just want to completely ban calling this on a type-checked argument list, what do you think?

Reply:
With the original arguments update this becomes `isTrailingClosureIndex`. Personally I prefer that over `isTrailingClosureArgument`, as that sounds like it would expect an `Argument` parameter, but I'm happy to change it over